### PR TITLE
[vNext Tokens] Adding a new Appearance editor

### PR DESCRIPTION
**Pull request by mischreiber**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

This PR adds a new view, `DemoAppearanceView`, along with some necessary accoutrements, to enable greater testing of custom Fluent design tokens.

The core component, `DemoAppearanceView`, is a SwiftUI view that features four controls:
* **App Color Scheme**: allows you to toggle between light and dark modes on demand.
* **Theme**: Acts the same as the theme picker in the `Colors` demo controller. This can be used to enable or disable the demo app's Green theme.
* **Theme-wide override**: Registers a set of custom design tokens for the current control type across the entire current Fluent theme.
* **Per-control override**: Individually sets `overrideTokens` for each Fluent control in the current demo screen.

The first two controls, **App Color Scheme** and **Theme**, are universal and function across the entire app by default. The latter two, **Theme-wide override** and **Per-control override**, require additional hookup by each demo screen to provide the relevant custom tokens. This PR includes sample implementations for `ButtonDemoController` and `CardNudgeDemoController`; additional demos will be implemented at a later time.

Finally, a note about **Theme-wide override**: this toggle does not immediately take effect for visible controls, due to the way that Fluent controls observe tokens from the current `FluentTheme`. Discussions as to whether this behavior will be changed will happen soon; any fallout from these discussions will be addressed in future PRs.

### Verification

https://user-images.githubusercontent.com/4934719/156511567-b4e6575c-bd09-4cb2-9f85-ea0ed7e31497.mov

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [x] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/879)

### CODE REVIEW

File: ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift [hunk 1]
```diff
@@ -158,4 +158,40 @@ class DemoController: UIViewController {
...
+                                    onPerControlOverrideChanged: @escaping ((Bool) -> Void)) {
+
+        appearanceController.popoverPresentationController?.delegate = self
+
+        // Store the callbacks from the individual demo controller
+        appearanceController.setupPerDemoCallbacks(onThemeWideOverrideChanged: onThemeWideOverrideChanged,
+                                                   onPerControlOverrideChanged: onPerControlOverrideChanged)
+
+        // Display the DemoAppearancePopover button
+        navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(systemName: "gear"),
```
Comment (huwilkes) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift:174]: Not sure it matters or how related they really are to this Fluent, but do we have a Fluent Icon for a gear? And would we want to use that?

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceController.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+/// Wrapper class to allow presenting of `DemoAppearanceView` from a UIKit host.
+class DemoAppearanceController: UIHostingController<DemoAppearanceView> {
+    init() {
+        let configuration = DemoAppearanceView.Configuration()
+        self.configuration = configuration
+
+        super.init(rootView: DemoAppearanceView(configuration: configuration))
+    }
+
+    @MainActor @objc required dynamic init?(coder aDecoder: NSCoder) {
```
Comment (rdeassis) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceController.swift:19]: Why do we need @MainActor / @objc / dynamic?

Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceController.swift:19]: Ha, no idea. Xcode wrote this line of code for me 🥇 

(I'll remove the things that I don't understand and that don't actually seem necessary.)

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 3]
```diff
@@ -0,0 +1,160 @@
+//
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import FluentUI
+import SwiftUI
+
+struct DemoAppearanceView: View {
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:9]: Excellent question! I'd like to implement a new base class for all of the `UIHostingController` instances we have so far and put the show/hide logic there. I hadn't decided if that base would itself inherit from `ControlHostingContainer`, but I'm leaning yes.

I kept it out of this PR to keep the size down, but it's something I can easily ("easily") add now so we have a complete story.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+    @ObservedObject var configuration: Configuration
+
+    /// Picker for setting the app's color scheme.
+    @ViewBuilder
+    var appColorSchemePicker: some View {
+        Text("App Color Scheme")
+            .font(.headline)
+        Picker("Color Scheme", selection: $configuration.colorScheme) {
+            Text("Light").tag(ColorScheme.light)
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:21]: Great idea!

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+            themePicker
+            FluentDivider()
+                .padding()
+
+            Text("Control")
+                .font(.headline)
+
+            // Theme-wide override toggle
+            Toggle(isOn: $configuration.themeWideOverride) {
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:55]: Oh yeah I forgot we had that thing. Will do!

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                Text("Theme-wide override")
+            }
+            .disabled(configuration.onThemeWideOverrideChanged == nil)
+
+            // Per-control override toggle
+            Toggle(isOn: $configuration.perControlOverride) {
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:61]: That's correct. The view controller is destroyed and rebuilt on navigation, so I'm not sure how we would persist that. Do you think we should?

File: ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift (hunk 1 above)
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift:174]: Hmm great point! We don't have one currently added to any asset catalogs, but that's probably something that can be changed. Do we have an official approach for how to add Fluent images to either FluentUI itself or the demo app specifically?

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift [hunk 7]
```diff
@@ -113,6 +107,47 @@ class ButtonDemoController: UITableViewController {
...
+
+        } onPerControlOverrideChanged: { [weak self] perControlOverrideEnabled in
+            let tokens = perControlOverrideEnabled ? self?.perControlOverrideTokens : nil
+            self?.buttons.forEach({ (_: String, value: MSFButton) in
+                value.state.overrideTokens = tokens
+            })
+        }
+    }
+
+    private var themeWideOverrideTokens: ButtonTokens {
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift:135]: Ah yeah, I called that out in the summary for this PR:

> Finally, a note about **Theme-wide override**: this toggle does not immediately take effect for visible controls, due to the way that Fluent controls observe tokens from the current `FluentTheme`. Discussions as to whether this behavior will be changed will happen soon; any fallout from these discussions will be addressed in future PRs.

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift [hunk 8]
```diff
@@ -113,6 +107,47 @@ class ButtonDemoController: UITableViewController {
...
+        return ButtonTokens(style: .primary,
+                            size: .large,
+                            textFont: .init(name: "Times",
+                                            size: 20.0,
+                                            weight: .regular))
+    }
+
+    private var perControlOverrideTokens: ButtonTokens {
+        return ButtonTokens(style: .primary,
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift:144]: Yep! Still tracked, and still something I intend to tackle soon.

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift [hunk 9]
```diff
@@ -31,6 +31,8 @@ class CardNudgeDemoController: DemoTableViewController {
                                                selector: #selector(updateCardNudgeSize),
                                                name: UIContentSizeCategory.didChangeNotification,
                                                object: nil)
+
+        self.configureAppearancePopover()
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift:35]: Goal was for this PR to be the proving ground for the change. If we're happy with the overall design then I'm happy to convert those too!

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift [hunk 10]
```diff
@@ -88,6 +90,41 @@ class CardNudgeDemoController: DemoTableViewController {
...
+
+        } onPerControlOverrideChanged: { [weak self] perControlOverrideEnabled in
+            let tokens = perControlOverrideEnabled ? self?.perControlOverrideTokens : nil
+            self?.cardNudges.forEach({ cardNudge in
+                cardNudge.state.overrideTokens = tokens
+            })
+        }
+    }
+
+    private var themeWideOverrideTokens: CardNudgeTokens {
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift:118]: Yep, that's the same issue as called out here:

> Finally, a note about **Theme-wide override**: this toggle does not immediately take effect for visible controls, due to the way that Fluent controls observe tokens from the current `FluentTheme`. Discussions as to whether this behavior will be changed will happen soon; any fallout from these discussions will be addressed in future PRs.

Fixing that will require some tweaks to FluentTheme that land outside the scope of this PR. But hey, our test code shook out a potential issue, so yay for test code!

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift [hunk 11]
```diff
@@ -88,6 +90,41 @@ class CardNudgeDemoController: DemoTableViewController {
...
+        return CardNudgeTokens(style: .standard,
+                               backgroundColor: DynamicColor(light: ColorValue(0xFF0000)))
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift:120]: Sure! That's an awful teal anyway.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:21]: Scratch that, it doesn't look like there's a way to set that value in UIKit :(

File: ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift (hunk 1 above)
Comment (harrieshin) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift:174]: not a very systematic way :/  but we try to keep the filename the same as the https://github.com/microsoft/fluentui-system-icons repo so we can update it later if necessary.

Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift:174]: Ah thanks for the pointer! Turns out we already have `ic_fluent_settings_24_regular` in the demo app, so I can just use that.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+                .padding()
+
+            themePicker
+            FluentDivider()
+                .padding()
+
+            Text("Control")
+                .font(.headline)
+
+            // Theme-wide override toggle
```
Comment (rdeassis) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:54]: It seems like GitHub won't let me reply to this [comment](https://github.com/microsoft/fluentui-apple/pull/879#discussion_r792178036):

I think it'd be nice to have an API in the design token system to tell whether the tokens are overridden for a specific control in a given theme. That way we could set the state of the toggle when the UI is being created.

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/ButtonDemoController.swift]: While a solution to this problem remains elusive, I've added an alert to at least let people know what to do after switching this toggle.

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/CardNudgeDemoController.swift]: As mentioned, this is now explained via a little alert, while we try to figure out how to get theme-wide tokens to apply live.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceController.swift [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+//  Copyright (c) Microsoft Corporation. All rights reserved.
+//  Licensed under the MIT License.
+//
+
+import FluentUI
+import SwiftUI
+import UIKit
+
+protocol DemoAppearanceDelegate: NSObjectProtocol {
+    func themeWideOverrideDidChange(isOverrideEnabled: Bool)
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceController.swift:11]: nit: add more documentation on all of these protocols?

File: ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift [hunk 14]
```diff
@@ -8,7 +8,9 @@ import Foundation
 /// Base class for all Fluent control tokenization.
 open class ControlTokens: NSObject {
     required public override init() {}
+
+    public var globalTokens: GlobalTokens { fluentTheme.globalTokens }
```
Comment (harrieshin) [inline ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift:12]: size still reasonable?

File: ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json [hunk 15]
```diff
@@ -2,23 +2,15 @@
...
-      "scale" : "3x"
+      "idiom" : "universal"
     }
   ],
   "info" : {
     "author" : "xcode",
     "version" : 1
   },
   "properties" : {
+    "preserves-vector-representation" : true,
```
Comment (harrieshin) [inline ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json:13]: I forget... is this the latest guidance to use this?

**Review by harrieshin — APPROVED** [review]

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+        return configuration.onThemeWideOverrideChanged != nil && configuration.themeOverridePreviouslyApplied != nil
+    }
+
+    private var showPerControlOverrideToggle: Bool {
+        return configuration.onPerControlOverrideChanged != nil
+    }
+
+    // MARK: Helpers for accessing the surrounding UIKit environment.
+
+    private var firstWindow: UIWindow? { UIApplication.shared.windows.first }
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:164]: It is extremely fishy for multi-window, but we don't really have that in the demo app yet, do we?

File: ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift (hunk 14 above)
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift:12]: Good question, I'll check. But hopefully two vars isn't tooooo bad.

File: ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json (hunk 15 above)
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json:13]: I believe so? My understanding is that a single vector is quite a bit smaller than multiple bitmaps, and that the performance impact is now negligible. But I can ask around to verify.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift (hunk 16 above)
[the remark this answers is not in the document]
Comment (harrieshin) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:165]: our demo app running in iPad will have multiwindow support right?

File: ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json (hunk 15 above)
Comment (harrieshin) [reply ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json:13]: https://microsoft.github.io/apple-ux-guide/iOSImageFileFormat.html#app-size-impact

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift (hunk 16 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:165]: TIL!

Well that's a bit more of a problem then...

File: ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift [hunk 17]
```diff
@@ -157,5 +157,35 @@ class DemoController: UIViewController {
...
+    }
+
+    // MARK: - Demo Appearance Popover
+
+    func configureAppearancePopover() {
+        // If a subclass implements `DemoAppearanceDelegate`, becocontrolTokensHashme the delegate.
+        appearanceController.delegate = self as? DemoAppearanceDelegate
+
+        // Display the DemoAppearancePopover button
+        navigationItem.rightBarButtonItem = UIBarButtonItem(image: UIImage(named: "ic_fluent_settings_24_regular"),
```
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoController.swift:168]: You're right, that's an old video! I've uploaded a new version that shows the correct icons.

File: ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json (hunk 15 above)
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/Resources/Assets.xcassets/LeftNav/ic_fluent_settings_24_regular.imageset/Contents.json]: Huh! Okay, reverted back to bitmap only. Thanks for catching that!

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift (hunk 16 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift]: Resolved! I simply moved most of the logic into the `UIHostingController` subclass, where it has better visibility into the current window. Check out the awesome multi-window demo video!

File: ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift (hunk 14 above)
Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/ControlTokens.swift:12]: Just checked: making these public adds 176 bytes to our binary, which I think I can live with!

File: ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (mischreiber) [reply ios/FluentUI.Demo/FluentUI.Demo/DemoAppearanceView.swift:11]: So there are some minor complications to getting things connected to SwiftUI. Let's chat offline next week about possible approaches, but for now I want to get this unblocked for validation of other controls.

**Pull request merged**